pageserver/src/tenant/storage_layer/merge_iterator.rs

```diff
@@ -0,0 +1,301 @@
+                (k1, l1, loaded_1).cmp(&(k2, l2, loaded_2))
+            }
+            (Some(_), None) => Ordering::Less,
+            (None, Some(_)) => Ordering::Greater,
+            (None, None) => Ordering::Equal,
+        }
+        .reverse()
+    }
+}
+
+impl<'a> IteratorWrapper<'a> {
+    pub fn create_from_image_layer(
+        image_layer: &'a ImageLayerInner,
+        ctx: &'a RequestContext,
+    ) -> Self {
+        Self {
+            peek_next_value: Some((
+                image_layer.key_range().start,
+                image_layer.lsn(),
+                Value::Image(Bytes::new()),
+            )),
+            layer: LayerRef::Image(image_layer),
+            ctx,
+            iter: None,
+        }
+    }
+
+    pub fn create_from_delta_layer(
+        delta_layer: &'a DeltaLayerInner,
+        ctx: &'a RequestContext,
+    ) -> Self {
+        Self {
+            peek_next_value: Some((
+                delta_layer.key_range().start,
+                delta_layer.lsn_range().start,
+                Value::Image(Bytes::new()),
+            )),
+            layer: LayerRef::Delta(delta_layer),
+            ctx,
+            iter: None,
+        }
+    }
+
+    fn peek_next_key_lsn(&self) -> Option<(&Key, Lsn)> {
+        let Some((key, lsn, _)) = &self.peek_next_value else {
+            return None;
+        };
+        Some((key, *lsn))
+    }
+
+    async fn load(&mut self) -> anyhow::Result<()> {
+        assert!(!self.is_loaded());
+        let mut iter = self.layer.iter(self.ctx);
+        self.peek_next_value = iter.next().await?;
+        self.iter = Some(iter);
+        Ok(())
+    }
+
+    fn is_loaded(&self) -> bool {
+        self.iter.is_some()
+    }
+
+    async fn next(&mut self) -> anyhow::Result<Option<(Key, Lsn, Value)>> {
+        if !self.is_loaded() {
+            self.load().await?;
+        }
+        let result = self.peek_next_value.take();
+        let iter = self.iter.as_mut().unwrap();
+        self.peek_next_value = iter.next().await?;
+        Ok(result)
+    }
+}
+
+pub struct MergeIterator<'a> {
+    heap: BinaryHeap<IteratorWrapper<'a>>,
+}
+
+impl<'a> MergeIterator<'a> {
+    pub fn create(
+        deltas: &[&'a DeltaLayerInner],
+        images: &[&'a ImageLayerInner],
+        ctx: &'a RequestContext,
+    ) -> Self {
+        let mut heap = BinaryHeap::with_capacity(images.len() + deltas.len());
+        for image in images {
+            heap.push(IteratorWrapper::create_from_image_layer(image, ctx));
+        }
+        for delta in deltas {
+            heap.push(IteratorWrapper::create_from_delta_layer(delta, ctx));
+        }
+        Self { heap }
+    }
+
+    pub async fn next(&mut self) -> anyhow::Result<Option<(Key, Lsn, Value)>> {
+        while let Some(mut iter) = self.heap.peek_mut() {
+            if !iter.is_loaded() {
+                iter.load().await?;
+            } else {
```

Comment:
Is this even needed? The `iter.next()` method has this exact logic internally.

Seems like a layering violation to me.

Reply:
removed iter.load in iter.next